skyd/skyd.go

```diff
@@ -1,16 +1,137 @@
+	// so we don't need to fetch that for each skylink we potentially want to
+	// pin/unpin.
+	pinnedSkylinksCache struct {
+		Skylinks   []string
+		Expiration time.Time
+		mu         sync.Mutex
 	}
 )
 
-/*
-TODO
- - env vars for talking to skyd - host, port, apipassword, etc.
- - methods for calling skyd:
-	- list all pinned skylinks
-	- pin a skylink
-	- unpin a skylink
-*/
+// NewClient creates a new skyd client.
+func NewClient(host, port, password string) *Client {
+	opts := client.Options{
+		Address:       fmt.Sprintf("%s:%s", host, port),
+		Password:      password,
+		UserAgent:     "Sia-Agent",
+		CheckRedirect: nil,
+	}
+	return &Client{
+		staticClient: client.New(opts),
+	}
+}
+
+// Pin instructs the local skyd to pin the given skylink.
+func (c *Client) Pin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if pinned {
+		// The skylink is already locally pinned, nothing to do.
+		return nil
+	}
+	sp, err := skymodules.NewSiaPath(skylink)
+	if err != nil {
+		return errors.AddContext(err, "failed to create siapath")
+	}
+	spp := skymodules.SkyfilePinParameters{
+		SiaPath: sp,
+	}
+	return c.staticClient.SkynetSkylinkPinPost(skylink, spp)
+}
+
+// PinnedSkylinks returns the list of skylinks pinned by the local skyd.
+func (c *Client) PinnedSkylinks() (skylinks []string, err error) {
+	skylinksCache.mu.Lock()
+	defer skylinksCache.mu.Unlock()
+	// Check whether the cache is still valid and return it if so.
+	if skylinksCache.Expiration.After(time.Now().UTC()) {
+		return skylinksCache.Skylinks, nil
+	}
+	// The cache is not valid, fetch the data from skyd.
+	sls := make([]string, 0)
+	dirsToWalk := []skymodules.SiaPath{skymodules.SkynetFolder}
+	for len(dirsToWalk) > 0 {
+		// Pop the first dir and walk it.
+		dir := dirsToWalk[0]
+		dirsToWalk = dirsToWalk[1:]
+
+		rd, err := c.staticClient.RenterDirRootGet(dir)
+		if err != nil {
+			return nil, errors.AddContext(err, "failed to fetch skynet directories from skyd")
+		}
+		for _, f := range rd.Files {
+			sls = append(sls, f.Skylinks...)
+		}
+		// Grab all subdirs and queue them for walking.
+		for _, d := range rd.Directories {
+			dirsToWalk = append(dirsToWalk, d.SiaPath)
+		}
+	}
+	// Update the cache.
+	skylinksCache.Skylinks = sls
+	skylinksCache.Expiration = time.Now().UTC().Add(skylinksCacheDuration)
+	return sls, nil
+}
+
+// Unpin instructs the local skyd to unpin the given skylink.
+func (c *Client) Unpin(skylink string) error {
+	pinned, err := c.isPinned(skylink)
+	if err != nil {
+		return err
+	}
+	if !pinned {
+		// The skylink is not locally pinned, nothing to do.
+		return nil
+	}
+	return c.staticClient.SkynetSkylinkUnpinPost(skylink)
+}
+
+// isPinned checks the list of skylinks pinned by the local skyd for the given
+// skylink and returns true if it finds it.
+func (c *Client) isPinned(skylink string) (bool, error) {
+	sls, err := c.PinnedSkylinks()
```

Comment:
This is a valid concern - the cache is using a lot or memory. The largest server we have currently is pinning a little under a million files, which takes about 100MB of memory if stored in a slice and about 200MB if stored in a map. This is not a trivial amount of memory to be throwing around for nothing.

Options:

1. Call `skyd` on each pin request and ask it if it's pinning the given skylink.

We can ignore unpin requests becase for those we can just call skyd with an unpin instruction and handle the error in the case it's not pinning the skylink. But the pin requests are going to be vastly more than the unpin requests and this call to skyd will add a lot of latency. We also don't have an endpoint like that in skyd and it seems that we don't have an efficient way of doing it. So, this is a dead end.

2. Tell `skyd` to pin the skylink each time without checking.

This will pin a new copy of the skylink because `skyd` won't check whether it's already pinning it for the same reasons as in 1 - there isn't a good way to do that. This is incredibly wasteful and it also isn't faster.

3. Keep a cached copy of all skylinks `skyd` has <TBC>

Reply:
> I'm not sure if `2` is not an option because we have lazy pinning now.

I don't think lazy pinning is checking whether the skylink is already locally pinned, either. So, it will still duplicate the data each time we pin. We can't have that.

Reply:
we currently are pulling servers from production if they are over 1.5 million files. So we have a process in place for limiting this memory overhead fwiw

So skyd will have scaling issues before this has scaling issues. 

So if this solution works for even up to 2 million skylinks, I think we can close this and come back to it when it is an issue.

Reply:
With 2 mil skylinks this cache consumes 405MB which I believe is acceptable.